File: bend2/effs/tcp_listen.js
// TCP
// ===

// TCP.listen owns its port; TCP.listen_shared (`shared`) sets SO_REUSEPORT
// first, so the listeners that all asked for it share the port. The C twin
// says why a refused option is the call's error.
function tcp_listen_with(port, shared) {
  const sys = io_sys();
  const fd = sys.socket(2, 1, 0);
  if (fd < 0) {
    return io_fail(sys.errno());
  }
  const one = new Int32Array([1]);
  const level = sys.mac ? 0xffff : 1;
  sys.setsockopt(fd, level, sys.mac ? 4 : 2, sys.ptr(one), 4);
  const reuse = sys.mac ? 0x200 : 15;
  if (shared && sys.setsockopt(fd, level, reuse, sys.ptr(one), 4) < 0) {
    const code = sys.errno();
    sys.close(fd);
    return io_fail(code);
  }
  const at = io_addr("0.0.0.0", Number(port));
  if (at === null) {
    sys.close(fd);
    return io_fail(22);
  }
  if (sys.bind(fd, sys.ptr(at), 16) < 0 || sys.listen(fd, 16) < 0
    || sys.fcntl(fd, 4, sys.fcntl(fd, 3, 0) | (sys.mac ? 4 : 0x800)) < 0) {
    const code = sys.errno();
    sys.close(fd);
    return io_fail(code);
  }
  return io_done(fd);
}

function tcp_listen(port) {
  return tcp_listen_with(port, false);
}

function tcp_listen_shared(port) {
  return tcp_listen_with(port, true);
}

File: bend2/effs/tcp_listen.c
// TCP
// ===

static int reuse_port(int fd) {
#ifdef SO_REUSEPORT
  int one = 1;
  return setsockopt(fd, SOL_SOCKET, SO_REUSEPORT, &one, sizeof(one));
#else
  errno = ENOPROTOOPT;
  return -1;
#endif
}

// TCP.listen owns its port: a second listener on it fails with EADDRINUSE,
// which is how an accidental second instance, or a probe for a free port,
// finds out. TCP.listen_shared (`shared`) sets SO_REUSEPORT first, so the
// listeners that all asked for it share the port, and a plain TCP.listen on
// that port still fails.
uint32_t tcp_listen(uint32_t port, int shared, int* out) {
  int fd = socket(AF_INET, SOCK_STREAM, 0);
  if (fd < 0) {
    return (uint32_t)errno;
  }
  int one = 1;
  setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));
  // Before the bind, or the kernel has already refused the second one. A
  // refusal is the call's error, not a silent downgrade to an owned port.
  // On Linux the kernel spreads accepted connections over the listeners;
  // on macOS it permits the bind but the first listener takes them all.
  if (shared && reuse_port(fd) < 0) {
    uint32_t code = (uint32_t)errno;
    close(fd);
    return code;
  }
  struct sockaddr_in at;
  if (io_sys_addr("0.0.0.0", port, &at) < 0) {
    close(fd);
    return EINVAL;
  }
  int bound = bind(fd, (struct sockaddr*)&at, sizeof(at));
  if (bound < 0 || listen(fd, 16) < 0
    || fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK) < 0) {
    uint32_t code = (uint32_t)errno;
    close(fd);
    return code;
  }
  *out = fd;
  return 0;
}

static Term tcp_listen_pack(Env e, uint32_t q, int out) {
  return q != 0 ? io_fail(e, q, NULL) : io_done(e, io_hand(out));
}

#ifdef CID_TCP_LISTEN
Term tcp_listen_run(Env e, Term* f, IoWork* w) {
  int out;
  uint32_t q = tcp_listen((uint32_t)f[0], 0, &out);
  return tcp_listen_pack(e, q, out);
}

static void __attribute__((constructor)) tcp_listen_use(void) {
  io_eff(CID_TCP_LISTEN, tcp_listen_run, 0);
}
#endif

#ifdef CID_TCP_LISTEN_SHARED
Term tcp_listen_shared_run(Env e, Term* f, IoWork* w) {
  int out;
  uint32_t q = tcp_listen((uint32_t)f[0], 1, &out);
  return tcp_listen_pack(e, q, out);
}

static void __attribute__((constructor)) tcp_listen_shared_use(void) {
  io_eff(CID_TCP_LISTEN_SHARED, tcp_listen_shared_run, 0);
}
#endif
